### PR TITLE
Render DAG and rendering modules architecture (updated)

### PR DESCRIPTION
**Pull request by 4Denthusiast**
An update of #3732 to fix the merge conflicts. Goes with https://github.com/Terasology/BasicRendering/pull/2 . All the original work was by @dave2s. I haven't even reviewed or tested the PR, except that it doesn't immediately crash and things are still getting rendered.

### CODE REVIEW

**Comment by pollend** [comment]
@4Denthusiast I can't get the game to run correctly. I see there are a couple things going on in this PR could we split some of this up into something more manageable.

**Comment by pollend** [comment]
there seems to be some outstanding problems with this. I can't seem to load a world correctly.

```
11:54:03.382 [main] INFO  o.t.w.b.internal.BlockManagerImpl - Registered Block CoreBlocks:Torch.BACK with id 65
11:54:07.151 [Thread-3] DEBUG c.j.d.entities.pipe.UnixPipe - Received packet: Pkt:FRAME{"evt":null,"data":{"assets":{"large_image":"515746954660151306"},"timestamps":{"start":1588186442000},"name":"Terasology","details":"Name: Player14144","state":"Solo | Game 4","secrets":{},"application_id":"515274721080639504"},"cmd":"SET_ACTIVITY","nonce":"95143052-10ff-49e3-97ff-8c1054afcec6"}
11:54:08.458 [main] ERROR o.terasology.engine.TerasologyEngine - Uncaught exception, attempting clean game shutdown
java.lang.NullPointerException: null
	at org.terasology.rendering.world.RenderableWorldImpl.isChunkVisible(RenderableWorldImpl.java:397)
	at org.terasology.rendering.world.RenderableWorldImpl.isChunkVisibleFromMainLight(RenderableWorldImpl.java:389)
	at org.terasology.rendering.world.RenderableWorldImpl.queueVisibleChunks(RenderableWorldImpl.java:313)
	at org.terasology.rendering.world.WorldRendererImpl.preRenderUpdate(WorldRendererImpl.java:334)
	at org.terasology.rendering.world.WorldRendererImpl.render(WorldRendererImpl.java:364)
	at org.terasology.engine.modes.StateIngame.render(StateIngame.java:206)
	at org.terasology.engine.subsystem.lwjgl.LwjglGraphics.postUpdate(LwjglGraphics.java:166)
	at org.terasology.engine.TerasologyEngine.tick(TerasologyEngine.java:473)
	at org.terasology.engine.TerasologyEngine.mainLoop(TerasologyEngine.java:427)
	at org.terasology.engine.TerasologyEngine.run(TerasologyEngine.java:403)
	at org.terasology.engine.Terasology.main(Terasology.java:162)
11:54:08.458 [main] INFO  o.terasology.engine.TerasologyEngine - Shutting down Terasology...
11:54:08.586 [main] INFO  o.t.p.i.ReadWriteStorageManager - Saving - Creating game snapshot
11:54:08.603 [main] INFO  o.t.p.i.ReadWriteStorageManager - Saving - Snapshot created: Writing phase starts
11:54:08.610 [main] INFO  o.t.n.internal.NetworkSystemImpl - Client disconnected: Player14144
11:54:08.624 [main] INFO  o.t.logic.console.ConsoleImpl - [NOTIFICATION] Player "Player14144" has left the game
11:54:08.625 [main] INFO  o.t.n.internal.NetworkSystemImpl - Network shutdown
11:54:08.656 [Thread-3] DEBUG c.j.d.entities.pipe.UnixPipe - Received packet: Pkt:FRAME{"evt":null,"data":{"assets":{"large_image":"515746954660151306"},"timestamps":{"start":1588186448000},"name":"Terasology","details":"Name: Player14144","state":"In Lobby","secrets":{},"application_id":"515274721080639504"},"cmd":"SET_ACTIVITY","nonce":"e2e6f153-a33a-4e32-aee6-7aa98c855b3e"}
11:54:08.746 [Engine-Task-Pool] INFO  o.t.rendering.opengl.ScreenGrabber - Screenshot saved to saves/Game 4/previews/1.jpg! 
11:54:09.096 [Saving-0] INFO  o.t.p.internal.SaveTransaction - Save game finished
11:54:09.100 [main] DEBUG c.j.d.entities.pipe.UnixPipe - Closing IPC pipe...
java.lang.NullPointerException
	at org.terasology.rendering.world.RenderableWorldImpl.isChunkVisible(RenderableWorldImpl.java:397)
	at org.terasology.rendering.world.RenderableWorldImpl.isChunkVisibleFromMainLight(RenderableWorldImpl.java:389)
	at org.terasology.rendering.world.RenderableWorldImpl.queueVisibleChunks(RenderableWorldImpl.java:313)
	at org.terasology.rendering.world.WorldRendererImpl.preRenderUpdate(WorldRendererImpl.java:334)
	at org.terasology.rendering.world.WorldRendererImpl.render(WorldRendererImpl.java:364)
	at org.terasology.engine.modes.StateIngame.render(StateIngame.java:206)
	at org.terasology.engine.subsystem.lwjgl.LwjglGraphics.postUpdate(LwjglGraphics.java:166)
	at org.terasology.engine.TerasologyEngine.tick(TerasologyEngine.java:473)
	at org.terasology.engine.TerasologyEngine.mainLoop(TerasologyEngine.java:427)
	at org.terasology.engine.TerasologyEngine.run(TerasologyEngine.java:403)
	at org.terasology.engine.Terasology.main(Terasology.java:162)
For more details, see the log files in /home/michaelpollind/drive/project/Terasology/logs/2020-04-29_11-53-50
```

**Comment by dave2s** [comment]
> there seems to be some outstanding problems with this. I can't seem to load a world correctly.
> 
> ```
> 11:54:03.382 [main] INFO  o.t.w.b.internal.BlockManagerImpl - Registered Block CoreBlocks:Torch.BACK with id 65
> 11:54:07.151 [Thread-3] DEBUG c.j.d.entities.pipe.UnixPipe - Received packet: Pkt:FRAME{"evt":null,"data":{"assets":{"large_image":"515746954660151306"},"timestamps":{"start":1588186442000},"name":"Terasology","details":"Name: Player14144","state":"Solo | Game 4","secrets":{},"application_id":"515274721080639504"},"cmd":"SET_ACTIVITY","nonce":"95143052-10ff-49e3-97ff-8c1054afcec6"}
> 11:54:08.458 [main] ERROR o.terasology.engine.TerasologyEngine - Uncaught exception, attempting clean game shutdown
> java.lang.NullPointerException: null
> 	at org.terasology.rendering.world.RenderableWorldImpl.isChunkVisible(RenderableWorldImpl.java:397)
> 	at org.terasology.rendering.world.RenderableWorldImpl.isChunkVisibleFromMainLight(RenderableWorldImpl.java:389)
> 	at org.terasology.rendering.world.RenderableWorldImpl.queueVisibleChunks(RenderableWorldImpl.java:313)
> 	at org.terasology.rendering.world.WorldRendererImpl.preRenderUpdate(WorldRendererImpl.java:334)
> 	at org.terasology.rendering.world.WorldRendererImpl.render(WorldRendererImpl.java:364)
> 	at org.terasology.engine.modes.StateIngame.render(StateIngame.java:206)
> 	at org.terasology.engine.subsystem.lwjgl.LwjglGraphics.postUpdate(LwjglGraphics.java:166)
> 	at org.terasology.engine.TerasologyEngine.tick(TerasologyEngine.java:473)
> 	at org.terasology.engine.TerasologyEngine.mainLoop(TerasologyEngine.java:427)
> 	at org.terasology.engine.TerasologyEngine.run(TerasologyEngine.java:403)
> 	at org.terasology.engine.Terasology.main(Terasology.java:162)
> 11:54:08.458 [main] INFO  o.terasology.engine.TerasologyEngine - Shutting down Terasology...
> 11:54:08.586 [main] INFO  o.t.p.i.ReadWriteStorageManager - Saving - Creating game snapshot
> 11:54:08.603 [main] INFO  o.t.p.i.ReadWriteStorageManager - Saving - Snapshot created: Writing phase starts
> 11:54:08.610 [main] INFO  o.t.n.internal.NetworkSystemImpl - Client disconnected: Player14144
> 11:54:08.624 [main] INFO  o.t.logic.console.ConsoleImpl - [NOTIFICATION] Player "Player14144" has left the game
> 11:54:08.625 [main] INFO  o.t.n.internal.NetworkSystemImpl - Network shutdown
> 11:54:08.656 [Thread-3] DEBUG c.j.d.entities.pipe.UnixPipe - Received packet: Pkt:FRAME{"evt":null,"data":{"assets":{"large_image":"515746954660151306"},"timestamps":{"start":1588186448000},"name":"Terasology","details":"Name: Player14144","state":"In Lobby","secrets":{},"application_id":"515274721080639504"},"cmd":"SET_ACTIVITY","nonce":"e2e6f153-a33a-4e32-aee6-7aa98c855b3e"}
> 11:54:08.746 [Engine-Task-Pool] INFO  o.t.rendering.opengl.ScreenGrabber - Screenshot saved to saves/Game 4/previews/1.jpg! 
> 11:54:09.096 [Saving-0] INFO  o.t.p.internal.SaveTransaction - Save game finished
> 11:54:09.100 [main] DEBUG c.j.d.entities.pipe.UnixPipe - Closing IPC pipe...
> java.lang.NullPointerException
> 	at org.terasology.rendering.world.RenderableWorldImpl.isChunkVisible(RenderableWorldImpl.java:397)
> 	at org.terasology.rendering.world.RenderableWorldImpl.isChunkVisibleFromMainLight(RenderableWorldImpl.java:389)
> 	at org.terasology.rendering.world.RenderableWorldImpl.queueVisibleChunks(RenderableWorldImpl.java:313)
> 	at org.terasology.rendering.world.WorldRendererImpl.preRenderUpdate(WorldRendererImpl.java:334)
> 	at org.terasology.rendering.world.WorldRendererImpl.render(WorldRendererImpl.java:364)
> 	at org.terasology.engine.modes.StateIngame.render(StateIngame.java:206)
> 	at org.terasology.engine.subsystem.lwjgl.LwjglGraphics.postUpdate(LwjglGraphics.java:166)
> 	at org.terasology.engine.TerasologyEngine.tick(TerasologyEngine.java:473)
> 	at org.terasology.engine.TerasologyEngine.mainLoop(TerasologyEngine.java:427)
> 	at org.terasology.engine.TerasologyEngine.run(TerasologyEngine.java:403)
> 	at org.terasology.engine.Terasology.main(Terasology.java:162)
> For more details, see the log files in /home/michaelpollind/drive/project/Terasology/logs/2020-04-29_11-53-50
> ```

You probably did not enable basic rendering module for the game. There is no check for this because no-one made the decision what to do in case this happens. And neither did anyone make a decision about which parts of the game should be made dependable on this since it's supposed to be optional.

**Comment by pollend** [comment]
that might be a bit of a problem if we want to the game to work. umm @Cervator or @skaldarnar have any idea with how to handle this in module space?

**Comment by Cervator** [comment]
Yep, have ideas 👍 Just yeah that's an area we haven't gotten around to dealing with.

Probably my ideal suggestion would be to introduce a line of module dependencies in the PC facade itself. It could be either just BasicRendering (and possible other crucial modules) or a full line of modules we consider fundamental to the game. Them being present as dependencies would mean Gradle would fetch them all and make them available, or if they're present in source form they should be guaranteed fully built as you run the game as a PC application - unlike one of the current issues where nothing from the engine/PC facade depends on modules therefore they might not be considered relevant to a build.

In any case that brings integrity to the workspace. You'll have the stuff you need. We'd also make sure whichever new modules remain in the Omega dist so they'll get shipped in the zip.

Secondly you can then enable the module as-needed for gameplay templates. By just making it a dependency of Core (like Health, Inventory, etc) you'll catch most cases. If BasicRendering or an alternate must _always_ be active we might need a new in-game dependency type, and again a set of fundamental modules _or alternatives thereof_ which will always be guaranteed active by the engine. Possibly only on clients - not sure if a headless server would need BasicRendering enabled, but the clients that connect would.

**Comment by 4Denthusiast** [comment]
Detecting the lack of a renderer during module dependency resolution would be nice, and should probably be done at some point, but it's much easier to detect the missing modules in WorldRendererImpl as there's a list of rendering modules there anyway, so I made it abort loading the world back to the main menu with an error message if there are no rendering modules loaded. Although there could be rendering modules designed as additions to others that don't actually offer enough support to stop the game crashing, in all those cases they should depend on another module anyway (like AdvancedRendering does on BasicRendering).

It still doesn't detect if AdvancedRendering is missing, but the fact that that causes a crash is just a bug anyway, so it would be better to fix that than force AdvancedRendering to load.

File: engine/src/main/java/org/terasology/config/Config.java [hunk 1]
```diff
@@ -67,6 +68,7 @@
  * Terasology user config. Holds the various global configuration information that the user can modify. It can be saved
  * and loaded in a JSON format.
  */
+@API
```
Comment (4Denthusiast) [inline engine/src/main/java/org/terasology/config/Config.java:71]: This gives modules access to the path where the game is. That doesn't seem like that much of a security problem, but I'm not sure.

File: engine/src/main/java/org/terasology/rendering/nui/layers/mainMenu/StartPlayingScreen.java [hunk 2]
```diff
@@ -101,6 +102,16 @@ public void initialise() {
         WidgetUtil.trySubscribe(this, "mainMenu", button -> {
             getManager().pushScreen("engine:mainMenuScreen");
         });
+
+        WidgetUtil.trySubscribe(this, "renderingSettings", button -> {
```
Comment (4Denthusiast) [inline engine/src/main/java/org/terasology/rendering/nui/layers/mainMenu/StartPlayingScreen.java:106]: I don't know enough about the menu system to work out the actual reason quickly, but returning from the render module set up screen then returning again (from world preview to pregeneration) doesn't work correctly (it goes back to the render module setup instead).

File: engine/src/main/java/org/terasology/config/Config.java (hunk 1 above)
Comment (DarkWeird) [reply engine/src/main/java/org/terasology/config/Config.java:71]: Yep, this gives modules access to this class.
Rendering Modules needs it for read RenderingConfigs.

**Comment by skaldarnar** [comment]
Superseded by #3990 and #3994

**Pull request closed without merging**